### PR TITLE
[Linux] Fix bxt_timer_autostop for games that stop based on a multi_manager 

### LINKED ISSUE #63: CMultiManager::ManagerUse hook crashes on Linux

### PR DESCRIPTION
**Pull request by chinese-soup**
Fixes #63 (works around it).

Like we've discussed with @YaLTeR on the SourceRuns Discord:
Fix bxt_timer_autostop for games like Blue-Shift, GMC, etc. that end on a multi_manager being fired (`CMultiManager__ManagerThink` on Windows). On Linux instead use a `FireTargets` hook, which is what is called in `MultiManager_ManagerThink()` anyway:

```c++
void CMultiManager :: ManagerThink ( void )
{
	float	time;

	time = gpGlobals->time - m_startTime;
	while ( m_index < m_cTargets && m_flTargetDelay[ m_index ] <= time )
	{
		FireTargets( STRING( m_iTargetName[ m_index ] ), m_hActivator, this, USE_TOGGLE, 0 );
		m_index++;
	}
// [...] etc.
}
```

All the remaining code in this PR for the autostopping is just a direct copy-paste from the Win `ManagerThink`, except that in `FireTargets` there's also a check if the classname of the pCaller is actually a `multi_manager` since `FireTargets` is called by many other entities.

**Tested in Blue Shift:** https://user-images.githubusercontent.com/5108747/121822559-fc518780-cc9f-11eb-9f89-7278f775a356.mp4

### CODE REVIEW

File: BunnymodXT/modules/ServerDLL.cpp [hunk 1]
```diff
@@ -31,10 +31,10 @@ extern "C" void __cdecl _ZN11COFGeneWorm10DyingThinkEv(void* thisptr)
 }
 
 // https://github.com/YaLTeR/BunnymodXT/issues/63
```
Comment (YaLTeR) [inline BunnymodXT/modules/ServerDLL.cpp:33]: This comment can go since it'll be fixed.

File: BunnymodXT/modules/ServerDLL.cpp [hunk 2]
```diff
@@ -709,12 +709,12 @@ void ServerDLL::FindStuff()
...
-		// ORIG_CMultiManager__ManagerUse_Linux = reinterpret_cast<_CMultiManager__ManagerUse_Linux>(MemUtils::GetSymbolAddress(m_Handle, "_ZN13CMultiManager10ManagerUseEP11CBaseEntityS1_8USE_TYPEf"));
-		if (ORIG_CMultiManager__ManagerUse_Linux)
-			EngineDevMsg("[server dll] Found CMultiManager::ManagerUse [Linux] at %p.\n", ORIG_CMultiManager__ManagerUse_Linux);
+		// https://github.com/YaLTeR/BunnymodXT/issues/63 <- because of this issue FireTargets is hooked on Linux instead, which is what MM::Think calls anyway
+		ORIG_FireTargets_Linux = reinterpret_cast<_FireTargets_Linux>(MemUtils::GetSymbolAddress(m_Handle, "_Z11FireTargetsPKcP11CBaseEntityS2_8USE_TYPEf"));
+		if (ORIG_FireTargets_Linux)
+			EngineDevMsg("[server dll] Found FireTargets_Linux [Linux] at %p.\n", ORIG_FireTargets_Linux);
 		else {
-			EngineDevWarning("[server dll] Could not find CMultiManager::ManagerThink or CMultiManager::ManagerUse.\n");
+			EngineDevWarning("[server dll] Could not find FireTargets_Linux or CMultiManager::ManagerUse.\n");
```
Comment (YaLTeR) [inline BunnymodXT/modules/ServerDLL.cpp:717]:
```suggestion
			EngineDevWarning("[server dll] Could not find FireTargets or CMultiManager::ManagerUse.\n");
```

File: BunnymodXT/modules/ServerDLL.hpp [hunk 3]
```diff
@@ -7,6 +7,7 @@
 
 class ServerDLL : public IHookableDirFilter
 {
+
```
Comment (YaLTeR) [inline BunnymodXT/modules/ServerDLL.hpp:10]: ye

File: BunnymodXT/modules/ServerDLL.cpp [hunk 4]
```diff
@@ -1365,29 +1365,48 @@ HOOK_DEF_2(ServerDLL, void, __fastcall, CMultiManager__ManagerThink, void*, this
...
-						CustomHud::SaveTimeToDemo();
-						RuntimeData::Add(RuntimeData::GameEndMarker {});
-					}
+			const char *classname = (*ppGlobals)->pStringBase + pev->classname;
+			const char *gameDir = "";
+			// We first need to check if the pCaller is a multi_manager since FireTargets can be called by anyone
+			if (!std::strcmp(classname, "multi_manager")) {
+				if (ClientDLL::GetInstance().pEngfuncs)
+					gameDir = ClientDLL::GetInstance().pEngfuncs->pfnGetGameDirectory();
+				if (!std::strcmp(targetname, "roll_the_credits")
```
Comment (YaLTeR) [inline BunnymodXT/modules/ServerDLL.cpp:1381]: Could you extract all of this into 1 function so there's no duplication?

**Review by YaLTeR — CHANGES_REQUESTED** [review]

File: BunnymodXT/modules/ServerDLL.hpp (hunk 3 above)
Comment (chinese-soup) [reply BunnymodXT/modules/ServerDLL.hpp:10]: ye

File: BunnymodXT/modules/ServerDLL.cpp (hunk 4 above)
Comment (chinese-soup) [reply BunnymodXT/modules/ServerDLL.cpp]: done

**Comment by chinese-soup** [comment]
Should be all fixed up.

File: BunnymodXT/modules/ServerDLL.cpp [hunk 5]
```diff
@@ -1365,29 +1345,55 @@ HOOK_DEF_2(ServerDLL, void, __fastcall, CMultiManager__ManagerThink, void*, this
...
+			const char *classname = (*ppGlobals)->pStringBase + pev->classname;
+			const char *gameDir = "";
+			// We first need to check if the pCaller is a multi_manager since FireTargets can be called by anyone
+			if (!std::strcmp(classname, "multi_manager")) {
+				if(IsMultiManagerAutoStopTargetName(targetname)) {
+					if (CVars::bxt_timer_autostop.GetBool())
+						CustomHud::SetCountingTime(false);
+					Interprocess::WriteGameEnd(CustomHud::GetTime());
+					CustomHud::SaveTimeToDemo();
+					RuntimeData::Add(RuntimeData::GameEndMarker{});
```
Comment (YaLTeR) [inline BunnymodXT/modules/ServerDLL.cpp:1364]: maybe move all of this there too and call it something like `DoMultiManagerTimerAutoStop`

File: BunnymodXT/modules/ServerDLL.cpp [hunk 6]
```diff
@@ -1365,29 +1345,55 @@ HOOK_DEF_2(ServerDLL, void, __fastcall, CMultiManager__ManagerThink, void*, this
...
+	    || (!std::strcmp(targetname, "multimanager_1") && !std::strcmp(gameDir, "construction")) // Construction
+	    || (!std::strcmp(targetname, "the_endgame_mm") && !std::strcmp(gameDir, "gloom")) // The Gloom
+	    || (!std::strcmp(targetname, "endbox_mm0") && !std::strcmp(gameDir, "echoes")) // Echoes
+	    || (!std::strcmp(targetname, "sendmm") && !std::strcmp(gameDir, "MINIMICUS"))  // Minimicus
+	    || (!std::strcmp(targetname, "kill2") && !std::strcmp(gameDir, "before")) // Before
+	    || (!std::strcmp(targetname, "tele_in") && !std::strcmp(gameDir, "plague")) // Plague
+	    || (!std::strcmp(targetname, "exit_seq") && !std::strcmp(gameDir, "timeline2")) // Timeline 2
+	    || (!std::strcmp(targetname, "spawn_garg_sci_mm") && !std::strcmp(gameDir, "SteamLink")) // Uplink
+	    || (!std::strcmp(targetname, "medicosprey") && !std::strcmp(gameDir, "visitors"))); // Visitors
+
```
Comment (YaLTeR) [inline BunnymodXT/modules/ServerDLL.cpp:1396]: 🙂

File: BunnymodXT/modules/ServerDLL.cpp (hunk 5 above)
Comment (YaLTeR) [reply BunnymodXT/modules/ServerDLL.cpp]: no I mean together with the check so it's one function

Comment (chinese-soup) [reply BunnymodXT/modules/ServerDLL.cpp]: alright, done

File: BunnymodXT/modules/ServerDLL.cpp [hunk 7]
```diff
@@ -1333,61 +1332,61 @@ HOOK_DEF_2(ServerDLL, void, __fastcall, CMultiManager__ManagerThink, void*, this
...
-					if (!std::strcmp(callerTargetname, "mgr_take_over") || !std::strcmp(callerTargetname, "endbot")) {
-						if (CVars::bxt_timer_autostop.GetBool())
-							CustomHud::SetCountingTime(false);
-						Interprocess::WriteGameEnd(CustomHud::GetTime());
-						CustomHud::SaveTimeToDemo();
-						RuntimeData::Add(RuntimeData::GameEndMarker {});
-					}
-				}
+			const char *classname = (*ppGlobals)->pStringBase + pev->classname;
+			const char *gameDir = "";
```
Comment (YaLTeR) [inline BunnymodXT/modules/ServerDLL.cpp:1350]:
```suggestion
```

**Comment by YaLTeR** [comment]
Thanks!

**Pull request merged**